hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/predicate/HazelcastComparisonPredicateUtils.java

```diff
@@ -121,6 +121,7 @@ private static boolean checkOperandTypesWithPrecedence(
         }
 
         if (highHZType.getTypeFamily() != lowHZType.getTypeFamily()
+            && !(highHZType.getTypeFamily().isTemporal() && lowHZType.getTypeFamily().isTemporal())
```

Comment:
It seems that proper type checking is missing here. If both types are temporal, we should check if the type with lower precedence could be cast to a type with higher precedence (e.g. `DATE` cannot be cast to `TIME`). If this is not the case, we should throw a signature error (`callBinding.newValidationSignatureError()`).